File: cmd/rillan/serve.go
package main

import (
	"errors"
	"log/slog"
	"os"

	"github.com/sidekickos/rillan/internal/app"
	"github.com/sidekickos/rillan/internal/config"
	"github.com/spf13/cobra"
)

func newServeCommand() *cobra.Command {
	var configPath string

	cmd := &cobra.Command{
		Use:   "serve",
		Short: "Start the local Rillan API daemon",
		RunE: func(cmd *cobra.Command, args []string) error {
			cfg, err := config.Load(configPath)
			if err != nil {
				return err
			}

			projectConfigPath := config.DefaultProjectConfigPath(cfg.Index.Root)
			projectCfg, err := config.LoadProject(projectConfigPath)
			if err != nil {
				if errors.Is(err, os.ErrNotExist) {
					projectCfg = config.DefaultProjectConfig()
				} else {
					return err
				}
			}

			systemConfigPath := config.DefaultSystemConfigPath()
			var systemCfg *config.SystemConfig
			loadedSystemCfg, err := config.LoadSystem(systemConfigPath)
			if err != nil {
				if !errors.Is(err, os.ErrNotExist) {
					return err
				}
			} else {
				systemCfg = &loadedSystemCfg
			}

			logger := newLogger(cfg.Server.LogLevel)
			application, err := app.New(cfg, projectCfg, systemCfg, configPath, projectConfigPath, systemConfigPath, logger)
			if err != nil {
				return err
			}

			return application.Run(cmd.Context())
		},
	}

	cmd.Flags().StringVar(&configPath, "config", config.DefaultConfigPath(), "Path to the runtime config file")

	return cmd
}

func newLogger(level string) *slog.Logger {
	logLevel := new(slog.LevelVar)
	logLevel.Set(config.ParseLogLevel(level))

	return slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: logLevel}))
}

File: crates/rillan/src/commands/serve.rs
// SPDX-FileCopyrightText: 2026 Rillan AI LLC
// SPDX-License-Identifier: Apache-2.0

//! `rillan serve` — start the local API daemon.

use std::path::PathBuf;

use anyhow::{Context, Result};
use clap::Args;
use rillan_app::App;
use rillan_secretstore::Store;

#[derive(Debug, Args)]
pub(crate) struct ServeArgs {
    /// Path to the runtime config file.
    #[arg(long, value_name = "PATH")]
    pub(crate) config: Option<PathBuf>,
}

pub(crate) async fn run(args: ServeArgs, store: Store) -> Result<()> {
    let config_path = args
        .config
        .unwrap_or_else(rillan_config::default_config_path);
    let cfg = rillan_config::load(&config_path)
        .with_context(|| format!("load config: {}", config_path.display()))?;

    let project_path = rillan_config::resolve_project_config_path(&cfg.index.root);
    let project_cfg = match rillan_config::load_project(&project_path) {
        Ok(cfg) => cfg,
        Err(rillan_config::Error::Read(io)) if io.kind() == std::io::ErrorKind::NotFound => {
            rillan_config::default_project_config()
        }
        Err(err) => return Err(err.into()),
    };

    let system_path = rillan_config::resolve_system_config_path();
    let system_cfg = match rillan_config::load_system(&system_path) {
        Ok(cfg) => Some(cfg),
        Err(rillan_config::Error::Read(io)) if io.kind() == std::io::ErrorKind::NotFound => None,
        Err(err) => return Err(err.into()),
    };

    let app = App::new(
        cfg,
        project_cfg,
        system_cfg,
        store,
        config_path,
        project_path,
        system_path,
    )
    .await?;

    let shutdown = async {
        if let Err(err) = tokio::signal::ctrl_c().await {
            tracing::error!(error = %err, "ctrl-c handler failed");
        }
    };
    app.run(shutdown).await?;
    Ok(())
}
